README.md
# Sacrina Node.js SDK

## System Requirements

- Node.js - https://nodejs.org/en/download/


## Installation

Install Sacrina SDK using npm:
	
	npm install sacrinasdk --save
	
Alternatively, clone this repository and use SDK locally:

	git clone https://github.com/sshussain270/Sacrina-node-SDK.git

Then go into the Sacrina SDK folder and run `npm link`:
	
	cd /Sacrina-node-SDK
	npm link
	
Then go inside the directory where you want to use SacrinaSDK, and run:

	npm link sacrinasdk


## Configuration

Register for Sacrina account and get your key. The use:


	var SacrinaSDK = require('sacrinasdk').SacrinaSDK;

    
    sacrinasdk = new SacrinaSDK();
    sacrinasdk.add_key('[your API key]');

## Usage

### Add dataset

You can add dataset as a list as:

    lis = ['data 1', 'data 2', 'data 3',..];
    sacrinasdk.add_dataset(list);
    
Or you can add data one by one as:

    sacrinasdk.add_data('Data 1');

### Upload dataset

Upload your dataset to Sacrina using:

    sacrinasdk.upload_dataset();
    
### Select an existing dataset

You can select an exisiting dataset using the id parameter:

    sacrinasdk.select_dataset([dataset_id]);

### Create model

To create a new model using your added dataset, use the following:

    sacrinasdk.create_model();

### Select model

To select an existing model using its id, use the following:

    sacrinasdk.select_model([model_id]);

### Train model

To start your model training, use the following:

    sacrinasdk.train_model();

### Check model status

To check the status of training, use the following:

    sacrinasdk.check_model_status();

### Create optional feature

To create a new optional feature, use the following:

    sacrinasdk.create_optional_feature([keywords]);

### Add optional feature sample

To add a sample to optional feature, use the following:

    sacrinasdk.add_optional_feature_sample([optional_feature_id], [title], [content]);

### Select optional feature

To select an optional feature, use the following

    sacrinasdk.select_optional_feature([optional_feature_id]);

### Extract optional feature

To extract optional feature, use the following:

    sacrinasdk.extract_optional_feature();

### Check optional feature status

To optional feature status, use the following:

    sacrinasdk.check_optional_feature_status();
### Create project

To create a new project, use the following:

    sacrinasdk.create_project(gen_min, gen_max, sector_min, sector_max, limit, optional_feature_ids);

### Select project

To select an existing project using its id, use the following:

    sacrinasdk.select_project(project_id);

### Execute project

To execute the selected project, use the following:

    sacrinasdk.execute_project();

### Check project status

To check the project status, use the following:

    sacrinasdk.check_project_status();

### Download results

To download the results of the selected project, use the following:

    sacrinasdk.download_results();


CHANGELOG.md
# CHANGELOG

## 1.0.5 - 2018-02-21
* Optional features added

## 1.0.4 - 2018-01-10
* README.md added

## 1.0.2 - 2018-01-09
* Minor changes, some names changed

## 1.0.1 - 2018-01-09
* Changed name to sacrinasdk


## 1.0.0 - 2018-01-09
* Initial release for Sacrina Node SDK


lib/sacrinarest.js
//Using Node.js and ES6

//You'll have to install this dependency : npm install request --save
var request = require('request');
class sacrinarest {
    // Use Sacrina REST API

    constructor() {

        this.key = null;
        this.dataset = [];
        this.headers = {};
        this.dataset_id = null;
        this.model_id = null;
        this.project_id = null;
        this.results = [];
        this.model_progress = 0;
        this.project_progress = 0;
        this.results_downloaded = false;
    }

    //add key
    add_key(key) {

        this.key = key
        this.headers ={
            'Accept': 'application/json',
            'Api-Key': this.key,
        };
    }

    //create dataset
    add_data(input_string) {

        if (typeof input_string !== "string") {

            console.log('input_string paramter is not string, please check your input');
        }
        else {

            this.dataset.push(input_string);
        }
    }

    add_dataset(input_array) {

        var notString = false;
        for(var value of input_array) {
            if(typeof value !== "string")
                notString = true;
        }
        if ((!(input_array instanceof Array)) || (notString == true)) {

            console.log('input_array paramter is not an array of strings, please check your input');
        }
         else {

            this.dataset.push.apply(this.dataset, input_array);
         }
    }

    //upload dataset to sacrina;

    upload_dataset() {

        //create dataset
        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/datasets/',
            headers : this.headers,
            json : {'title':'my-dataset','description':'asdasdasdasd'}
        };
        request.post(options,function(error, response, body) {
        if (error || response.statusCode != 201) {
            console.log('error creating dataset');
        }
        
        that.dataset_id = body['id'];
         for (var item of that.dataset) {
            var options = {
                strictSSL : false,
                url : 'https://sacrina.com/REST/learning/datas/',
                headers : that.headers,
                json : {'title':'data1', 'dataset_id': 'https://sacrina.com/REST/learning/datasets/' 
                + String(that.dataset_id) + '/', 'content': String(item) }
            }
            request.post(options,function(error, response, body) {
                if (error || response.statusCode != 201) {
                    console.log('error uploading data to dataset');
                }
               
            });
        }
         console.log('Dataset upload completed') ;
        });

        //upload data to dataset
       
        
    }

    //add an existing dataset
    select_dataset(dataset_id){

        this.dataset_id = dataset_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/datasets/' + String(this.dataset_id) + '/',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }

    create_model(){

        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/',
            headers : this.headers,
            json : {'title':'mymodel','datasets': 'https://sacrina.com/REST/learning/datasets/' + String(this.dataset_id) + '/'}
        };
        
        request.post(options,function(error, response, body) {
            
            if (error || response.statusCode != 201)
                console.log("error");

            else {

                var response_json = body;
                that.model_id = response_json['id'];
                console.log(body);
            }
        });
    }

    //add an old model
    select_model(model_id) {

        this.model_id = model_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/',
            headers : this.headers
        };
         request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }

    //train the model
    train_model(){

        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/?train',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {
            console.log("TRENING: ");
            console.log(body);
            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }


    //check model training status
    check_model_status() {

        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else {
                
                var response_json = JSON.parse(body);
                var status = response_json['status'];
                that.model_progress = response_json['progress'];
                console.log(that.model_progress);
            }
        });
    }

    //create a project with the model
    create_project(gen, sector_min, sector_max) {

        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/',
            headers : this.headers,
            json : {'name':'myproject','model': 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/', 'gen': gen, 'sector_min': sector_min, 'sector_max': sector_max }
        };
        
        request.post(options,function(error, response, body) {
            if (response.statusCode != 201)
                return "error";

            else {
                var response_json = body;
                this.project_id = response_json['id'];
                return body;
            }
        });
    }

    //add an existing project
    select_project(project_id) {

        this.project_id = project_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/',
            headers : this.headers
        };
       
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else 
                return body;
        });

    }

    //execute the project
    execute_project() {
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/?execute',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else 
                return body;
        });

    }

    //check project status
    check_project_status() {
         var that = this;
         var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/',
            headers : this.headers
        };
        
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                console.log("error");
            else 
                var response_json = JSON.parse(body);
                var status = response_json['status'];
                that.project_progress = response_json['progress'];
                console.log(that.project_progress);
        });

    }

    //download results
    download_results() {
        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/results/?project_id=' + String(this.project_id),
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                console.log("error");
            else {
                var response_json = body;
                for (var item of response_json){
                    if (response_json.hasOwnProperty(key) && key == 'content')      
                            that.results.push(item["content"]);
                      }
                }
                that.results_downloaded = true;

            
        });
    }
    
}
exports.sacrinarest = sacrinarest;


lib/sacrinasdk.js
//Using Node.js and ES6

//You'll have to install this dependency like this: npm install request --save
var request = require('request');
class SacrinaSDK {
    // Use Sacrina REST API

    constructor() {

        this.key = null;
        this.dataset = [];
        this.headers = {};
        this.dataset_id = null;
        this.model_id = null;
        this.project_id = null;
        this.results = [];
        this.model_progress = 0;
        this.project_progress = 0;
        this.results_downloaded = false;
    }

    //add key
    add_key(key) {

        this.key = key
        this.headers ={
            'Accept': 'application/json',
            'Api-Key': this.key,
        };
    }

    //create dataset
    add_data(input_string) {

        if (typeof input_string !== "string") {

            console.log('input_string paramter is not string, please check your input');
        }
        else {

            this.dataset.push(input_string);
        }
    }

    add_dataset(input_array) {

        var notString = false;
        for(var value of input_array) {
            if(typeof value !== "string")
                notString = true;
        }
        if ((!(input_array instanceof Array)) || (notString == true)) {

            console.log('input_array paramter is not an array of strings, please check your input');
        }
         else {

            this.dataset.push.apply(this.dataset, input_array);
         }
    }

    //upload dataset to sacrina;

    upload_dataset() {

        //create dataset
        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/datasets/',
            headers : this.headers,
            json : {'title':'my-dataset','description':'asdasdasdasd'}
        };
        request.post(options,function(error, response, body) {
        if (error || response.statusCode != 201) {
            console.log('error creating dataset');
        }
        
        that.dataset_id = body['id'];
         for (var item of that.dataset) {
            var options = {
                strictSSL : false,
                url : 'https://sacrina.com/REST/learning/datas/',
                headers : that.headers,
                json : {'title':'data1', 'dataset_id': 'https://sacrina.com/REST/learning/datasets/' 
                + String(that.dataset_id) + '/', 'content': String(item) }
            }
            request.post(options,function(error, response, body) {
                if (error || response.statusCode != 201) {
                    console.log('error uploading data to dataset');
                }
               
            });
        }
         console.log('Dataset upload completed') ;
        });

        //upload data to dataset
       
        
    }

    //add an existing dataset
    select_dataset(dataset_id){

        this.dataset_id = dataset_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/datasets/' + String(this.dataset_id) + '/',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }

    create_model(){

        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/',
            headers : this.headers,
            json : {'title':'mymodel','datasets': 'https://sacrina.com/REST/learning/datasets/' + String(this.dataset_id) + '/'}
        };
        
        request.post(options,function(error, response, body) {
            
            if (error || response.statusCode != 201)
                console.log("error");

            else {

                var response_json = body;
                that.model_id = response_json['id'];
                console.log(body);
            }
        });
    }

    //add an old model
    select_model(model_id) {

        this.model_id = model_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/',
            headers : this.headers
        };
         request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }

    //train the model
    train_model(){

        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/?train',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {
            console.log("TRENING: ");
            console.log(body);
            if (error || response.statusCode != 200)
                return "error";
            else
                return body;
        });
    }


    //check model training status
    check_model_status() {

        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else {
                
                var response_json = JSON.parse(body);
                var status = response_json['status'];
                that.model_progress = response_json['progress'];
                console.log(that.model_progress);
            }
        });
    }

    //create a project with the model
    create_project(gen, sector_min, sector_max) {

        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/',
            headers : this.headers,
            json : {'name':'myproject','model': 'https://sacrina.com/REST/learning/models/' + String(this.model_id) + '/', 'gen': gen, 'sector_min': sector_min, 'sector_max': sector_max }
        };
        
        request.post(options,function(error, response, body) {
            if (response.statusCode != 201)
                return "error";

            else {
                var response_json = body;
                this.project_id = response_json['id'];
                return body;
            }
        });
    }

    //add an existing project
    select_project(project_id) {

        this.project_id = project_id;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/',
            headers : this.headers
        };
       
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else 
                return body;
        });

    }

    //execute the project
    execute_project() {
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/?execute',
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                return "error";
            else 
                return body;
        });

    }

    //check project status
    check_project_status() {
         var that = this;
         var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/projects/' + String(this.project_id) + '/',
            headers : this.headers
        };
        
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                console.log("error");
            else 
                var response_json = JSON.parse(body);
                var status = response_json['status'];
                that.project_progress = response_json['progress'];
                console.log(that.project_progress);
        });

    }

    //download results
    download_results() {
        var that = this;
        var options = {
            strictSSL : false,
            url : 'https://sacrina.com/REST/production/results/?project_id=' + String(this.project_id),
            headers : this.headers
        };
        request.get(options,function(error, response, body) {

            if (error || response.statusCode != 200)
                console.log("error");
            else {
                var response_json = body;
                for (var item of response_json){
                    if (response_json.hasOwnProperty(key) && key == 'content')      
                            that.results.push(item["content"]);
                      }
                }
                that.results_downloaded = true;

            
        });
    }
    
}
exports.SacrinaSDK = SacrinaSDK;
